### DISCUSSION TITLE
Mohon bantu saya untuk menampilkan tabel hasil subsetting di Streamlit

### DISCUSSION THREAD

**Discussion by ammaralwandi97** [teamalgoritma/community#822] [Q&A] [answered]
Pada intinya saya ingin menghasilkan Tabel Hasil filterisasi hasil input slider tapi saya gatau codenya mohon dibantu.
Ini pickelnya https://drive.google.com/file/d/1YuYpISv3j8N6xgjO1CDV9_STcbHAB3jk/view?usp=sharing

Berikut codingan py saya :

`import pandas as pd
import plotly.express as px
import streamlit as st

st.set_page_config(layout='wide')

# --- READ DATA ---
mitra = pd.read_pickle('data/mitra.pkl')

# --- ROW 1 ---
st.write('# Dasbor Mitra BPS Kota Sungai Penuh')
st.write("""Mitra Statistik adalah ujung tombak kualitas data statistik. Seiring dengan meningkatnya 
         kualitas dan kuantitas survei, maka diperlukan dasbor Mitra Statistik BPS Kota Sungai Penuh 
         untuk merekrut Mitra terbaik untuk pekerjaan Sensus dan Survei di BPS""")

st.divider()
col3, col4 = st.columns(2)

## --- INPUT SELECT ---
input_select = col3.selectbox(
    label='Pilih Kabupaten/Kota Asal Pendaftar',
    options=mitra['city'].unique()
)

## --- INPUT SLIDER ---
input_slider = col4.slider(
    label='Pilih Rentang Usia',
    min_value=mitra['age'].min(),
    max_value=mitra['age'].max(),
    value=[20,50]
)

min_slider = input_slider[0]
max_slider = input_slider[1]

# --- ROW 4 ---
col5, col6 = st.columns(2)

## --- BARPLOT ---
# data: barplot
mitra = mitra[mitra['city'] == input_select]
df_kec = pd.crosstab(index=mitra.district, columns='Total', colnames=[None])
df_kec = df_kec.reset_index()

# plot: barplot
plot_gen = px.bar(df_kec.sort_values(by='Total',ascending=False), x='district', y='Total', 
                   labels = {'district' : 'Kecamatan',
                             'Total' : 'Jumlah Mitra'})

col5.write(f'### Jumlah Mitra dengan KTP {input_select}  ') # f-string
col5.plotly_chart(plot_gen, use_container_width=True)

## --- MULTIVARIATE ---
# data: multivariate
employ_age = mitra[mitra['age'].between(left=min_slider, right=max_slider)]
dept_gender = pd.crosstab(index=employ_age['Pendidikan'],
                          columns=employ_age['Jenis Kelamin'],
                          colnames=[None])
dept_gender_melt = dept_gender.melt(ignore_index=False, var_name='Jenis Kelamin', value_name='num_people')
dept_gender_melt = dept_gender_melt.reset_index()

# plot: multivariate
plot_dept = px.bar(dept_gender_melt.sort_values(by='num_people'), 
                   x="num_people", y="Pendidikan", 
                   color="Jenis Kelamin", 
                   barmode='group',
                   labels = {'num_people' : 'Jumlah Mitra',
                             'Pendidikan' : 'Pendidikan',
                             'Jenis Kelamin': 'Jenis Kelamin'}
                             )

col6.write(f'### Jumlah Mitra KTP {input_select} Menurut Pendidikan dan Jenis Kelamin, Umur {min_slider} s/d {max_slider} Tahun')
col6.plotly_chart(plot_dept, use_container_width=True)

# --- ROW 7 ---
col6, col7 = st.columns(2)

input_slider_ujian = col7.slider(
    label='Pilih Rentang Nilai',
    min_value=0,
    max_value=100,
    value= [70,95]
)

min_ujian_slider = input_slider_ujian[0]
max_ujian_slider = input_slider_ujian[1]

nilai_gen = mitra.pivot_table(index='generation',values='Nilai Ujian',aggfunc='mean').round(2)\
    .sort_values(by='Nilai Ujian')
nilai_gen_melt = nilai_gen.reset_index()
plot_nilai_gen = px.bar(nilai_gen_melt, x='generation',y='Nilai Ujian',
       barmode='group',
                   labels = {'generation' : 'Generasi',
                             'Nilai Ujian' : 'Nilai Ujian'}
                             )
col6.write(f'### Rata-rata Nilai Ujian Mitra KTP {input_select} Menurut Generasi') # f-string
col6.plotly_chart(plot_nilai_gen, use_container_width=True)

filter_mitra = mitra[(mitra['Nilai Ujian'] >= min_ujian_slider) & 
                           (mitra['Nilai Ujian'] <= max_ujian_slider)]
col7.write(f'### Daftar Mitra KTP {input_select} Terpilih')
col7.table(filter_mitra)`

**Comment by divakartika** [comment]
Dear @ammaralwandi97 ,

Dapat diinformasikan kolom apa yang ingin difilter menggunakan input slider? Jika ingin filter kolom `age` maka cara yang terlampir sudah benar. 

Kami sarankan Pak Ammar untuk membuat plot di dalam file `.ipynb` terlebih dahulu untuk mengecek apakah plot yang ingin dihasilkan sudah sesuai, setelah itu code nya dapat dipindahkan ke `streamlit_app.py` dan disambungkan ke input slider.

Boleh dikirimkan kembali gambar / screenshot plot yang telah dibuat di `.ipynb`, atau gambar contoh plot yang ingin dihasilkan, terima kasih.

**Reply by ammaralwandi97** [reply]
minta contoh code untuk menampilkan tabel apapun ke streamlit. yang kemaren di contohkan menampilkan plot. menampilkan tabel belum, hehehe

**Reply by divakartika** [reply, marked as answer]
Untuk menampilkan tabel, bisa menggunakan `st.dataframe(nama_dataframe)` dengan `nama_dataframe` adalah data/tabel yang ingin ditampilkan. 

Untuk menampilkan dataframe pada salah satu kolom, misalkan telah dibuat `col1, col2 = st.columns(2)` maka dapat dibuat `col1.dataframe(nama_dataframe)` untuk menampilkan tabel di sebelah kiri.

Referensi
- `st.dataframe()`: https://docs.streamlit.io/library/api-reference/data/st.dataframe
- Fungsi-fungsi pada streamlit: https://docs.streamlit.io/library/api-reference

Semoga membantu dan menjawab pertanyaan Anda :)

**Reply by ammaralwandi97** [reply]
Alhamdulillah terima kasih mba, ternyata ada data yg missing

**Reply by divakartika** [reply]
Dear @ammaralwandi97 

Kami telah mencoba menjalankan code yang dilampirkan dan menemui error berikut.

![image](https://github.com/teamalgoritma/community/assets/79572421/127b1b71-a1e4-4de2-b164-b1789df653f1)

Error tersebut terjadi karena terdapat salah satu kolom dengan tipe data `timedelta64[ns]`, yaitu tipe data yang dihasilkan dari selisih antara dua waktu. Tipe data ini baru saja di-support oleh streamlit dengan versi terbaru.

![image](https://github.com/teamalgoritma/community/assets/79572421/f67edd93-dd0f-4011-a8d8-87f5221d23d2)

Sebagai solusi, Pak Ammar dapat melakukan upgrade versi streamlit dengan `pip install streamlit==1.29.0` pada terminal Visual Studio Code. Jangan lupa untuk melakukan `conda activate dss_nik` terlebih dahulu.

![image](https://github.com/teamalgoritma/community/assets/79572421/b1d338af-54a5-44d9-8eb9-513e81fe0fd5)

Selain itu versi streamlit pada `requirements.txt` juga perlu disesuaikan, untuk memberi tahu Streamlit Cloud versi streamlit yang perlu digunakan saat deployment

![image](https://github.com/teamalgoritma/community/assets/79572421/78004632-7f23-462c-bc7f-4d6f68506eac)

Silakan lakukan penyesuaian di atas pada file local computer dan **GitHub**. Panduan **update code GitHub** dapat dilihat pada [link berikut](https://docs.google.com/document/d/1HZG4ov_g_lKH0nQ3U1Ef2GKDYktvtFqEM77t7M6XG4Q/edit#heading=h.y55ofaz0808l).

Semoga membantu dan menjawab pertanyaan Anda :)
